docker/docker_image_dest.go

```diff
@@ -99,35 +108,50 @@ func (d *dockerImageDestination) PutBlob(digest string, stream io.Reader) error
 	logrus.Debugf("Uploading %s", uploadURL)
 	res, err = d.c.makeRequest("POST", uploadURL, nil, nil)
 	if err != nil {
-		return err
+		return "", -1, err
 	}
 	defer res.Body.Close()
 	if res.StatusCode != http.StatusAccepted {
 		logrus.Debugf("Error initiating layer upload, response %#v", *res)
-		return fmt.Errorf("Error initiating layer upload to %s, status %d", uploadURL, res.StatusCode)
+		return "", -1, fmt.Errorf("Error initiating layer upload to %s, status %d", uploadURL, res.StatusCode)
 	}
 	uploadLocation, err := res.Location()
 	if err != nil {
-		return fmt.Errorf("Error determining upload URL: %s", err.Error())
+		return "", -1, fmt.Errorf("Error determining upload URL: %s", err.Error())
 	}
 
 	// FIXME: DELETE uploadLocation on failure
 
+	h := sha256.New()
+	pr, pw := io.Pipe()
+	bufin := bufio.NewReader(stream)
+	mw := io.MultiWriter(h, pw)
+	go func() {
+		bufin.WriteTo(mw)
+		pw.Close()
+	}()
+	hash := h.Sum(nil)
```

Comment:
I can’t see how this works; by the time `h.Sum` is called here, the goroutine may not even have started.

Overall the only time a `sha256` can compute a digest is when the input stream has been read to the end; but we need the stream at the _beginning_ to send it to `http` (or, well, we could read it all to memory and run out of memory, or to disk and be slow; let’s not do either).